src/UserAgentMiddleware.php

```diff
@@ -0,0 +1,333 @@
+            return $executionEnvironment;
+        }
+
+        return "";
+    }
+
+    private function endpointDiscovery(): string
+    {
+        $args = $this->args;
+        if (isset($args['endpoint_discovery'])) {
+            if (($args['endpoint_discovery'] instanceof Configuration
+                && $args['endpoint_discovery']->isEnabled())
+            ) {
+                return 'cfg/endpoint-discovery';
+            } elseif (is_array($args['endpoint_discovery'])
+                && isset($args['endpoint_discovery']['enabled'])
+                && $args['endpoint_discovery']['enabled']
+            ) {
+                return 'cfg/endpoint-discovery';
+            }
+        }
+
+        return "";
+    }
+
+    /**
+     * Returns the user agent value for app id, but just when an
+     * app id was provided as a client argument.
+     *
+     * @return string
+     */
+    private function appId(): string
+    {
+        if (empty($this->args['app_id'])) {
+            return "";
+        }
+
+        return 'app/' . $this->args['app_id'];
+    }
+
+    /**
+     * Returns the user agent value for metrics.
+     *
+     * @return string
+     */
+    private function metrics(): string
+    {
+        static $metricsFn = [
+            'endpointMetric',
+            'accountIdModeMetric',
+            'retryConfigMetric'
+        ];
+        foreach ($metricsFn as $fn) {
+            $this->{$fn}();
+        }
+
+        $metricsEncoded = $this->metricsBuilder->build();
+        if (empty($metricsEncoded)) {
+            return "";
+        }
+
+        return "m/" . $metricsEncoded;
+    }
+
+    /**
+     * Appends the endpoint metric into the metrics builder,
+     * just if a custom endpoint was provided at client construction.
+     */
+    private function endpointMetric(): void
+    {
+        if (!empty($this->args['endpoint'])) {
+            $this->metricsBuilder->append(MetricsBuilder::ENDPOINT_OVERRIDE);
+        }
+    }
+
+    /**
+     * Appends the account id endpoint mode metric into the metrics builder,
+     * based on the account id endpoint mode provide as client argument.
+     */
+    private function accountIdModeMetric(): void
+    {
+        $accountIdMode = $this->args['account_id_endpoint_mode'] ?? null;
+        if ($accountIdMode === null) {
+            return;
+        }
+
+        if ($accountIdMode === 'preferred') {
+            $this->metricsBuilder->append(MetricsBuilder::ACCOUNT_ID_MODE_PREFERRED);
+        } elseif ($accountIdMode === 'disabled') {
+            $this->metricsBuilder->append(MetricsBuilder::ACCOUNT_ID_MODE_DISABLED);
+        } elseif ($accountIdMode === 'required') {
+            $this->metricsBuilder->append(MetricsBuilder::ACCOUNT_ID_MODE_REQUIRED);
+        }
+    }
+
+    /**
+     * Appends the retry mode metric into the metrics builder,
+     * based on the resolved retry config mode.
+     */
+    private function retryConfigMetric(): void
```

Comment:
Another case where I think we could use a static mapping

Reply:
maybe I'm missing something, but I think something like 

```php
$this->metricsBuilder->append(
    self::$staticMap[$retrymode]
);
```
could be done in this method without the need for additional functions